src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs

```diff
@@ -121,38 +126,64 @@ protected override void FormatInlineCount(IList<string> commandClauses)
         }
 
         private IList<string> SelectPathSegmentColumns(
-            IList<string> columns, string path, IList<string> excludePaths = null)
+            IList<string> columns, string path, EntityCollection collection, IList<string> excludePaths = null)
         {
             if (string.IsNullOrEmpty(path))
             {
                 var resultColumns = columns.Where(x => !HasMultipleSegments(x)).ToList();
                 if (excludePaths != null)
-                    resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
-                        !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                {
+                    if (excludePaths.Count == 1 && excludePaths[0] == "*")
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !_session.Metadata.GetNavigationPropertyNames(collection.Name).Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
```

Comment:
This is going to enumerate `_session.Metadata.GetNavigationPropertyNames(collection.Names)` roughly `columns.Count` times. You should hoist out the enumeration and `FormatFirstSegment` result into a set.

Reply:
Same as above..